AK/PrintfImplementation.h

```diff
@@ -283,166 +285,227 @@ ALWAYS_INLINE int print_signed_number(PutChFunc putch, char*& bufptr, int number
+    bool dot = false;
+    unsigned field_width = 0;
+    bool has_fraction_length = false;
+    unsigned fraction_length = 6;
+    unsigned long_qualifiers = 0;
+    bool size_qualifier = false;
+    bool alternate_form = 0;
+    bool always_sign = false;
+};
+
+template<typename PutChFunc, typename ArgumentListT, template<typename T, typename U = ArgumentListT> typename NextArgument>
+struct PrintfImpl {
+    ALWAYS_INLINE int format_s(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT ap) const
+    {
+        const char* sp = NextArgument<const char*>()(ap);
+        if (!sp)
+            sp = "(null)";
+        return print_string(putch, bufptr, sp, strlen(sp), state.left_pad, state.field_width, state.dot);
+    }
+    ALWAYS_INLINE int format_d(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        if (state.long_qualifiers >= 2)
+            return print_i64(putch, bufptr, NextArgument<i64>()(ap), state.left_pad, state.zero_pad, state.field_width);
+
+        return print_signed_number(putch, bufptr, NextArgument<int>()(ap), state.left_pad, state.zero_pad, state.field_width, state.always_sign);
+    }
+    ALWAYS_INLINE int format_i(PutChFunc putch, char*& bufptr, int nwritten, const State& state, ArgumentListT& ap) const
+    {
+        return format_d(putch, bufptr, nwritten, state, ap);
+    }
+    ALWAYS_INLINE int format_u(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        if (state.long_qualifiers >= 2)
+            return print_u64(putch, bufptr, NextArgument<u64>()(ap), state.left_pad, state.zero_pad, state.field_width);
+        return print_number(putch, bufptr, NextArgument<u32>()(ap), state.left_pad, state.zero_pad, state.field_width);
+    }
+    ALWAYS_INLINE int format_Q(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_u64(putch, bufptr, NextArgument<u64>()(ap), state.left_pad, state.zero_pad, state.field_width);
+    }
+    ALWAYS_INLINE int format_q(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<u64>()(ap), false, false, state.left_pad, state.zero_pad, 16);
+    }
+    ALWAYS_INLINE int format_g(PutChFunc putch, char*& bufptr, int nwritten, const State& state, ArgumentListT& ap) const
+    {
+        return format_f(putch, bufptr, nwritten, state, ap);
+    }
+    ALWAYS_INLINE int format_f(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_double(putch, bufptr, NextArgument<double>()(ap), state.left_pad, state.zero_pad, state.field_width, state.fraction_length);
+    }
+    ALWAYS_INLINE int format_o(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        if (state.alternate_form) {
+            putch(bufptr, '0');
+            return 1;
```

Comment:
???

Reply:
Ask whomever wrote that (@awesomekling, probably), I haven't changed any behaviour.

Reply:
That's `%#o`, which gives the "alternate" octal representation (zero-prefixed.)

Reply:
Sure, but why does it return at this point, instead of falling through to printing the actual value? As written, it just writes out a `'0'` character.

Reply:
oops.
Appears that is my mistake.

Reply:
it's `++ret` on master, my bad.